app/src/main/java/pm/gnosis/heimdall/ui/safe/details/transactions/SafeTransactionsViewModel.kt

```diff
@@ -31,44 +29,63 @@ class SafeTransactionsViewModel @Inject constructor(
+                    if (pending.isNotEmpty()) {
+                        combined += AdapterEntry.Header(R.string.header_pending)
+                        pending.forEach { combined += AdapterEntry.Transaction(it.id) }
+                    }
+                    var lastHeaderRes = 0
+                    submitted.forEach {
+                        val timeDiff = currentTime - it.timestamp
+                        val headerRes = when {
+                            timeDiff < DateTimeUtils.DAY_IN_MS -> R.string.header_today
+                            timeDiff < 2 * DateTimeUtils.DAY_IN_MS -> R.string.header_yesterday
+                            lastHeaderRes == 0 -> R.string.header_submitted // First header should never be "older"
+                            else -> R.string.header_older
+                        }
+                        if (lastHeaderRes != headerRes && headerRes != 0) {
+                            combined += AdapterEntry.Header(headerRes)
+                            lastHeaderRes = headerRes
+                        }
+                        combined += AdapterEntry.Transaction(it.id)
+                    }
+                    combined
+                }
+            )
+                .scanToAdapterData(initialData = cachedData)
+                .doOnNext { cachedData = it }
                 .mapToResult()
-        } ?: Flowable.empty<Result<Adapter.Data<String>>>()
+        } ?: Flowable.empty<Result<Adapter.Data<AdapterEntry>>>()
     }
 
-    override fun loadTransactionInfo(id: String): Single<Pair<TransactionInfo, TransferInfo?>> =
-        address?.let {
-            transactionInfoRepository.loadTransactionInfo(id)
-                .flatMap { info ->
-                    (info.data as? TransactionData.AssetTransfer)?.let {
-                        if (it.token == ERC20Token.ETHER_TOKEN.address) {
-                            // We always want to display the complete amount (all 18 decimals)
-                            val value = (Wei(it.amount)).toEther().stringWithNoTrailingZeroes()
-                            val symbol = context.getString(R.string.currency_eth)
-                            Single.just(info to TransferInfo(value, symbol))
-                        } else {
-                            loadTokenValue(it.token, it.amount).map { info to it.toNullable() }
-                        }
-                    } ?: Single.just(info to null)
+    override fun loadTransactionInfo(id: String): Single<TransactionInfo> =
+        transactionInfoRepository.loadTransactionInfo(id)
+
+    override fun loadTokenInfo(token: Solidity.Address): Single<ERC20Token> =
+        if (token == ERC20Token.ETHER_TOKEN.address)
+            Single.just(ERC20Token.ETHER_TOKEN)
+        else
+            tokenRepository.loadToken(token)
+                .onErrorResumeNext {
+                    tokenRepository.loadTokenInfo(token).firstOrError()
```

Comment:
It's a bit confusing that we have `loadToken` for something that is loaded locally and `loadTokenInfo` for a remote call 🤔

Reply:
hmm, we can rename that later :D has always bin like that. Maybe we should also refactor this into one method with a boolean that indicates if it should try fetching from the network. will change that with WA-420